heron/instance/src/java/org/apache/heron/instance/spout/SpoutInstance.java

```diff
@@ -296,9 +299,6 @@ public void run() {
           spoutMetrics.updateOutQueueFullCount();
         }
 
-        // Check if we have any message to process anyway
-        readTuplesAndExecute(streamInQueue);
-
```

Comment:
ping @nwangtw

Reply:
I made a small change to the message processing order for `SpoutInstance`.

Before, when the spout task is waken up by the event loop, it will attempt to emit tuples first. And then, it will process the control messages. Now it will process the control tuples first and then try to emit the tuples. Note that this change will only impact stateful topology (spout instances in non-stateful topology will not receive any control messages)

Implementing blocking emitting tuples until checkpoint is completed requires to process control messages first.

Reply:
This change should be fine. 

But I think you will need to update the `realTuplesAndExecute` method in the SpoutInstance to actually implement the blocking mechanism, just like what you did for the `BoltInstance`. Currently, the blocking is implemented in the `isProduceTuple` method

Reply:
`readTuplesAndExecute` in `SpoutInstance` only reads control tuples, and we don't need to block them. For spouts we only want to block `nextTuple` and for bolts we want to block `execute` (which takes place in the `readTuplesAndExecute` method).

Reply:
The ordering change sounds ok to me.

One concern is the barrier though. I think each stream needs a separated blocker. I just want to confirm that this is included in the PR.

Reply:
Chatted with Ning offline. He was worried about "barrier alignment" (eg. a task has multiple input streams). It is fine because "barrier alignment" is done at the stream manager 🙂

Reply:
Yeah. Good point. Should be safe then. Thanks~